x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export class SpacesSavedObjectsClient {
+  constructor(options) {
+    const {
+      request,
+      baseClient,
+      spaceUrlContext,
+    } = options;
+
+    this.errors = baseClient.errors;
+
+    this._client = baseClient;
+    this._request = request;
+
+    this._spaceUrlContext = spaceUrlContext;
+  }
+
+  async create(type, attributes = {}, options = {}) {
+
+    if (this._isTypeSpaceAware(type)) {
+      options.extraBodyProperties = {
+        ...options.extraBodyProperties,
+        spaceId: await this._getSpaceId()
+      };
+    }
+
+    return await this._client.create(type, attributes, options);
+  }
+
+  async bulkCreate(objects, options = {}) {
+    options.extraBodyProperties = {
+      ...options.extraBodyProperties,
+      spaceId: await this._getSpaceId()
+    };
+
+    return await this._client.bulkCreate(objects, options);
+  }
+
+  async delete(type, id) {
+    return await this._client.delete(type, id);
+  }
+
+  async find(options = {}) {
+    const spaceOptions = {};
+
+    if (this._isTypeSpaceAware(options.type)) {
+      const spaceId = await this._getSpaceId();
+
+      spaceOptions.queryDecorator = (query) => {
+        const { bool = {} } = query;
+
+        if (!Array.isArray(bool.filter)) {
+          bool.filter = [];
+        }
+
+        bool.filter.push({
+          term: {
+            spaceId
+          }
+        });
+
+        return query;
+      };
+    }
+
+    return await this._client.find({ ...options, ...spaceOptions });
+  }
+
+  async bulkGet(objects = []) {
+    // ES 'mget' does not support queries, so we have to filter results after the fact.
+    const thisSpaceId = await this._getSpaceId();
+
+    return await this._client.bulkGet(objects, {
```

Comment:
In a similar fashion to the above, how do you feel about passing in an `extraSourceProperties` array that we use to copy the additional properties from the document? Then we could just do the filter on the result without adding the extension point.

Reply:
I like it! Done!